src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,341 @@
+{
+    using LengthType                        = uint16_t;
+    static constexpr LengthType kNullLength = 0xFFFF;
+};
+
+// ember assumptions ... should just work
+static_assert(sizeof(ShortPascalString::LengthType) == 1);
+static_assert(sizeof(LongPascalString::LengthType) == 2);
+
+/// Encode the value stored in 'decoder' into an ember format string 'out'
+///
+/// The value encoded will be of type T (e.g. CharSpan or ByteSpan) and it will be encoded
+/// via the given ENCODING (i.e. ShortPascalString or LongPascalString)
+///
+/// isNullable defines if the value of NULL is allowed to be encoded.
+template <typename T, class ENCODING>
+CHIP_ERROR DecodeStringLikeIntoEmberBuffer(AttributeValueDecoder decoder, bool isNullable, MutableByteSpan out)
+{
+    T workingValue;
+
+    if (isNullable)
+    {
+        typename DataModel::Nullable<T> nullableWorkingValue;
+        ReturnErrorOnFailure(decoder.Decode(nullableWorkingValue));
+
+        if (nullableWorkingValue.IsNull())
+        {
+            VerifyOrReturnError(out.size() >= sizeof(typename ENCODING::LengthType), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            typename ENCODING::LengthType nullLength = ENCODING::kNullLength;
+            memcpy(out.data(), &nullLength, sizeof(nullLength));
+            return CHIP_NO_ERROR;
+        }
+
+        // continue encoding non-null value
+        workingValue = nullableWorkingValue.Value();
+    }
+    else
+    {
+        ReturnErrorOnFailure(decoder.Decode(workingValue));
+    }
+
+    typename ENCODING::LengthType len = static_cast<typename ENCODING::LengthType>(workingValue.size());
+    VerifyOrReturnError(out.size() >= sizeof(len) + len, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    memcpy(out.data(), &len, sizeof(len));
+    memcpy(out.data() + sizeof(len), workingValue.data(), workingValue.size());
+    return CHIP_NO_ERROR;
+}
+
+/// Encodes a numeric data value of type T from the given ember-encoded buffer `data`.
+///
+/// isNullable defines if the value of NULL is allowed to be encoded.
+template <typename T>
+CHIP_ERROR DecodeIntoEmberBuffer(AttributeValueDecoder & decoder, bool isNullable, MutableByteSpan out)
+{
+    if (isNullable)
+    {
+        using Traits = NumericAttributeTraits<T>;
+
+        DataModel::Nullable<typename Traits::WorkingType> workingValue;
+        ReturnErrorOnFailure(decoder.Decode(workingValue));
+
+        typename Traits::StorageType storageValue;
+        if (workingValue.IsNull())
+        {
+            Traits::SetNull(storageValue);
+        }
+        else
+        {
+            VerifyOrReturnError(Traits::CanRepresentValue(isNullable, *workingValue), CHIP_ERROR_INVALID_ARGUMENT);
+            Traits::WorkingToStorage(*workingValue, storageValue);
+        }
+
+        VerifyOrReturnError(out.size() >= sizeof(storageValue), CHIP_ERROR_INVALID_ARGUMENT);
+
+        const uint8_t * data = Traits::ToAttributeStoreRepresentation(storageValue);
+        memcpy(out.data(), data, sizeof(storageValue));
```

Comment:
Indicate byte ordering assumptions for this native encoding. Is it really native on all platforms?

Reply:
Updated. This is attribute store functions, so they take care of endianess..